src/gp_creds.c

```diff
@@ -667,6 +667,41 @@ uint32_t gp_add_krb5_creds(uint32_t *min,
             if (ret_maj) {
                 goto done;
             }
+
+            if (req_name != GSS_C_NO_NAME) {
+                int equal = 0;
+
+                ret_maj = gss_inquire_cred(&ret_min, impersonator_cred,
+                                           &target_name, NULL, NULL, NULL);
+                if (ret_maj) {
+                    goto done;
+                }
+
+                ret_maj = gss_compare_name(&ret_min, target_name,
+                                           req_name, &equal);
+                if (ret_maj) {
+                    goto done;
+                }
+
+                /* if impersonator credential retrieval yielded the requested client
+                 * name, we do not need to impersonate. Also, in AD an attempt to
+                 * impersonate oneself yields an error "KDC has no support for padata
+                 * type" */
+                if (equal) {
+                     ret_maj = gss_acquire_cred_from(&ret_min, req_name, GSS_C_INDEFINITE,
+                                                     &desired_mechs, cred_usage,
+                                                     &cred_store, &user_cred,
+                                                     actual_mechs, NULL);
```

Comment:
We actually do not need this call I think, given we already have a valid credential in impersonator_cred.
Unless you are concerned that cred_usage is CRED_BOTH instead of whatever the client really requested.

We can simply immediately return with *output_cred_handle = impersonator_cred I believe.

Reply:
GSS_C_BOTH is exactly what I was concerned about. I see a chance of handing a credential with more than the configured cred_usage to the client. Is that a possibility? On the other hand: Does the impersonator credential actually need to be GSS_C_BOTH? It's been a while since I read up on Constrained Delegation. During S4U2Self the ticket to self might actually be accepted against the impersonator credential for some reason. But maybe not. Would that be worth a try?

Reply:
The impersonator credentials must be GSS_C_BOTH, because otherwise MIT libraries deny impersonation (there are technical reasons for that, can't be changed).
I understand your hesitation and I think that it is ok to make another call just to be safe, avoid having to think what could possibly go wrong if the admin really wants to restrict creds to initiate type for some reason (present or future).
So ignore my comment to remove it, let's just fix how the return code is handled and then we can merge.